src/app/application.cpp

```diff
@@ -771,6 +795,64 @@ void Application::shutdownCleanup(QSessionManager &manager)
 }
 #endif
 
+#ifdef Q_OS_WIN
+namespace
+{
+    std::uint64_t autoMemoryWorkingSetSize(const std::uint64_t memorySize)
+    {
+        const std::uint64_t GB = 1024 * 1024 * 1024;
+
+        return (memorySize <= (4 * GB))
+                ? memorySize / 2
+                : (2 * GB) + (memorySize / 10);
+    }
+}
+
+void Application::applyMemoryWorkingSetLimit()
+{
+    MEMORYSTATUSEX statex;
+    statex.dwLength = sizeof(statex);
+    if (!::GlobalMemoryStatusEx(&statex))
+    {
+        const DWORD errorCode = ::GetLastError();
+        QString reason = tr("Error %1.").arg(errorCode);
+        LPVOID lpMsgBuf = nullptr;
+        if (::FormatMessageW((FORMAT_MESSAGE_ALLOCATE_BUFFER | FORMAT_MESSAGE_FROM_SYSTEM | FORMAT_MESSAGE_IGNORE_INSERTS)
+                         , nullptr, ::GetLastError(), MAKELANGID(LANG_NEUTRAL, SUBLANG_DEFAULT), reinterpret_cast<LPWSTR>(&lpMsgBuf), 0, nullptr))
+        {
+            reason = QString::fromWCharArray(reinterpret_cast<LPWSTR>(lpMsgBuf)).trimmed();
```

Comment:
This should not overwrite the error code but append after the code.

Reply:
> This should not overwrite the error code but append after the code.

Does it matter to the user? I just used it as a fallback in case I couldn't get the appropriate message. If you insist on showing the error code anyway, what format should it have? (so that I don't have to redo it multiple times)

Reply:
Let's move the discussion over here: https://github.com/qbittorrent/qBittorrent/pull/16485#discussion_r810796078